moveit_core/collision_distance_field/src/collision_distance_field_types.cpp

```diff
@@ -280,7 +280,7 @@ void collision_detection::BodyDecomposition::init(const std::vector<shapes::Shap
   bodies_.clear();
   for (unsigned int i = 0; i < shapes.size(); i++)
   {
-    bodies_.addBody(shapes[i]->clone(), poses[i], padding);
+    bodies_.addBody(shapes[i].get(), poses[i], padding);
```

Comment:
As the passed `shapes` are a _const_ vector of _const_ ShapePtrs, I'm tempted to assume that `clone()` was correct. It doesn't make sense to extract the raw pointer from the shared ptr and continue using the raw pointer, while the shared pointer might get destroyed...
I think the correct way of handling this is to
a) either store shared ptrs in `bodies_` as well, or
b) free the raw shape pointers before doing `bodies_.clear()`.

Reply:
Can you describe a situation where the shared_pts might get destroyed? What happens there, is that values from those Shape objects are used to construct (and add) new bodies but the shape objects themselves are not stored as part of bodies. That is, they are not required after the function call to init().

Reply:
Shouldn't the [destroctor/clear](https://github.com/ros-planning/geometric_shapes/blob/melodic-devel/src/bodies.cpp#L1311) functions take care of freeing the memory .?

Reply:
Sorry for the noise. I was thinking that `addBody(Shape*)` just stores a pointer to the passed shape (which then might end up in a dangling pointer). However, actually `addBody()` creates a copy internally:
https://github.com/ros-planning/geometric_shapes/blob/df0478870b8592ce789ee1919f3124058c4327d7/src/body_operations.cpp#L43-L66